### DISCUSSION TITLE
Can we use PSR-15 for Middlewares?

### DISCUSSION THREAD

**Discussion by leonhusmann** [clue/framework-x#130] [General]
Hi, thanks for working on this very interesting framework!

I like how you stick to [PSR-7](https://www.php-fig.org/psr/psr-7/) for your Requests and Responses. The community definitely benefits from those standards. That's why I wonder why you decided not to go for [PSR-15](https://www.php-fig.org/psr/psr-15/) for your Middlewares. Is there any technical reason or just personal preference?

Best regards, Leon

**Comment by SimonFrings** [comment]
@leonhusmann thanks for bringing this up :+1:

This is indeed something we have in mind for Framework X. Before we added our Fiber support we always had to return a promise instead of the actual value behind it. This is actually against PSR-15 which requires a `ResponseInterface` as the return value. With Fibers we're now awaiting these promises and get our actual results. So you opened this ticket at just the right time ^^

The implementation behind this is not that big, the most work will be the documentation.

I hope this gives you some insights into our plans :rocket:

**Comment by drupol** [comment]
What's the status of this request?

**Reply by SimonFrings** [reply]
@drupol, this feature is currently outstanding. However, we sat together as a team and came up with an overview of some of the outstanding topics in Framework X. You can check out our overview on Twitter: https://twitter.com/cassifyit/status/1718888717023334414

To keep you updated: We're currently involved in working on the ReactPHP MySQL connection pool, as discussed in https://github.com/friends-of-reactphp/mysql/issues/175. This is something we will benefit from in this project as well. After that, we want to give Framework X its first version, as we plan to introduce a few BC-Breaks in the future (move some files around, etc.). This way, we have proper versioning and can safely apply several changes. This will also include revisiting the PSR-15 feature at a later time.

Hope this gives you some insights in what's currently going on :+1:

**Reply by drupol** [reply]
Thank you.

**Comment by bwaidelich** [comment]
FYI: If you want to use existing PSR-15 Middlewares you could write a simple adapter like:

```php
final class MiddlewareAdapter {

    public function __construct(private readonly MiddlewareInterface $wrapped) {
    }

    public function __invoke(ServerRequestInterface $request, callable $next): ResponseInterface
    {
        $handler = new class(Closure::fromCallable($next)) implements RequestHandlerInterface {
            public function __construct(private readonly Closure $next) {}

            public function handle(ServerRequestInterface $request): ResponseInterface {
                return ($this->next)($request);
            }
        };
        return $this->wrapped->process($request, $handler);
    }
}
```

And use it like this:

```php
$app = new App(new MiddlewareAdapter($somePsr15Middleware));
```

It would be great if that work around wasn't required though :)

**Reply by SimonFrings** [reply]
@bwaidelich thanks for your input on this and sharing your `MiddlewareAdapter` :+1:

> It would be great if that work around wasn't required though :)

Agreed! As you might already know, we're currently working on the first version release for Framework X which should go live early next week (see https://twitter.com/x_framework/status/1762499173830778907). While working out the next upcoming features, we also went through all the prototypes on @clue's laptop and made a list of future topics we want to work. I can safely say that PSR-15 is one of those topics, so we might see this in an upcoming release not too far in the future :+1:

If you need this urgently, you can also reach out to us or help us prioritizing by [becoming as sponsor](https://github.com/sponsors/clue) ❤️

**Comment by bwaidelich** [comment]
I just realized, that this topic was brought up with https://github.com/clue/framework-x/discussions/185 too (maybe those threads can be merged?)

**Reply by SimonFrings** [reply]
I think I opened #185 because of this ticket, but it's been a while and I'm not 100% sure what my intention were :D

In my opinion #185 should be the place to discuss PSR-15, not sure tho how we would merge these two conversations. In my opinion it's good enough to set these references to other tickets :+1:
